rewrite autoPatchelfHook in python

layus opened the pull request:
###### Motivation for this change

Fixes #148547 

###### Things done

Reimplemented autoPatchelfHook in python for readability, proper cache dicts and speed!

/cc @aszlig sorry for pinging twice in a row. I would really appreciate your feedback on the implementation. For now it resides in a separate hook for ease of testing. It should replace the old one before merge.

In particular, the logic is now more clear, but maybe not the right one. And addAutoPatchelfSearchPath semantincs is broken, but works with all nixpkgs examples.

layus committed:
rewrite autoPatchelfHook in python

aszlig commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+            print("    {} -> found: {}".format(dep, foundDependency))
+        else:
+            print("    {} -> not found!".format(dep))
+            yield dep
```
Not personally a fan of this, since it does make the functions behaviour less obvious. Without knowing/looking at its implementation, what would you say `autoPatchelfFile` would yield or return?

layus replied:
Good point. And at the same time making that a global variable is not so nice either. Maybe it makes sense here.

aszlig replied:
I'd probably return a [`dataclass`](https://docs.python.org/3/library/dataclasses.html) or [`namedtuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple), eg. like this:

```python
@dataclass
class Dependency:
    path: Path
    found: bool

def auto_patchelf_file(path: Path, deps: List[Dependency]) -> List[Dependency]:
    ...
```

layus replied:
I used a global in the end. emphasis is on testing ATM. I also struggle with type annotations. What is the right way to check/enforce them ?

layus replied:
done :-). My global was not really global, and usign global (the keyword) is ugly and error prone.

aszlig reviewed (commented):
Thanks a lot for the effort! This btw. would probably also make it less ugly to use it standalone, extend it or even move it out of tree one day.

I added a few comments here and there, some of them are rather nit-picky and opinionated so feel free to ignore them.

Speaking of nit-picky: I'd also generally prefer to be in line with [PEP-8](https://www.python.org/dev/peps/pep-0008/), eg. using snake case instead of camelCase names for functions and variables. The reason is that it's more consistent with other Python code in nixpkgs and also matches that of most Python FOSS libraries out there.

Most of the comments also assume the use of `pathlib.Path` and I also usually tend use type annotations.

Anyway, as mentioned, these are just personal opinions.

If it comes to functionality: Have you actually tested this against software using `autoPatchelfHook` and checked whether it produces different results? Especially when it comes to `androidenv`, since the latter uses some of the provided shell functions we had so far.

layus commented on pkgs/build-support/setup-hooks/auto-patchelf-py.sh:
```diff
@@ -0,0 +1,91 @@
+# as we overwrite it ourselves for all the binaries).
+# BUT, changing phases breaks tools that rely on having a properly functioning
+# install after the fixupPhase.
+export dontPatchELF=1
```
It is not needed, but it is a sensible default as running the patchelfHook makes little sense. autoPatchelfHook is a strict superset of features.

aszlig replied:
If it's not needed, I would not set it because even if we leave aside the scenario where you want to set `dontAutoPatchelf` and use `autoPatchelf` manually, it also will cause surprising/confusing behaviour if you do eg. `stdenv.mkDerivation { ...; dontPatchelf = false; }`.

layus committed:
Update pkgs/build-support/setup-hooks/auto-patchelf.py

Co-authored-by: aszlig <aszlig@redmoonstudios.org>

layus committed:
Update pkgs/build-support/setup-hooks/auto-patchelf.py

Co-authored-by: aszlig <aszlig@redmoonstudios.org>

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+
+@contextmanager
+def openELF(path):
+    stream = open(path, "rb")
+    try:
+        elf = ELFFile(stream)
+        elf.filename = path
+        yield elf
+    finally:
+        stream.close()
```
Yes, I wanted this to be way more powerful, but it falled short.

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+    # is DYN.
+    return elf.header["e_type"] == 'ET_EXEC' or elf.get_section_by_name(".interp")
+
+def getDepsFromElfBinary(elf):
+    #if section := elf.get_section_by_name('.dynamic')
+    #    yield from enumerate_needed(section);
+    # does not always work ! (why ??)
+
+    for section in elf.iter_sections():
+        if isinstance(section, DynamicSection):
+            if section.name != ".dynamic":
+                print("Hu hoo: found you >{}<".format(section.name))
+            for tag in section.iter_tags('DT_NEEDED'):
+                yield tag.needed
+            # break ? There _should_ only be one. Not too sure anymore at this point.
+
+def getRpathFromElfBinary(elf):
+    if section := elf.get_section_by_name('.dynamic'):
+        for tag in section.iter_tags('DT_RUNPATH'):
+            return tag.runpath.split(':')
+
+        for tag in section.iter_tags('DT_RPATH'):
+            return tag.rpath.split(':')
+
+    return []
```
Not too sure about this one. The main reason that triggered this change is performance. As this runs on a lot of files and is supposed to fail when they are not ELFs calling patchelf on each may not be efficient. I have to test.

aszlig replied:
:+1: Let me know the results. Also, we could use [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) or even [asyncio](https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.create_subprocess_exec) for submitting the invocations into a worker pool if it's necessary.

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+
+        cachedPaths.add(lib)
+
+        for (root, dirs, files) in os.walk(lib):
+            used = False
+            for file in files:
+                if ".so" not in file:
+                    continue
+
```
Makes sense, with a switch for the non recursive glob at times.

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+
+                        if rpath := getRpathFromElfBinary(elf):
+                            libDirs += rpath
+                        #print("Caching", file, arch, lib, osabi)
```
A bit beyond my scope ATM. But a nice addition if we move it out of tree someday

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+
+    if interpreterArch != fileArch:
+        # Our target architecture is different than this file's architecture, so skip it.
+        print("skipping {} because its architecture ({}) differs from target ({})"
+            .format(elf.filename, fileArch, interpreterArch))
```
Oh, new python style ! I did not use python enough for a while :-)

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+            # just updating the rpath won't satisfy it).
+            if os.path.isfile(dep):
+                continue
+        elif os.path.isfile(os.path.join(libcLib, dep)):
```
Not quite, but I get the idea :-)

aszlig replied:
Oh, sorry, forgot the `/ dep` X-D

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+            print("    {} -> not found!".format(dep))
+            yield dep
+
+    # dedup. Relies on dicts preserving insertion order. Python3.7+
```
:-)

layus replied:
I guess I will keep at least the "dedup" part. Otherwise this arcane trick is not obvious.

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+    failed = False
+    for path in walk(pathsToPatch, recurse):
+        try:
+            with openELF(path) as elf:
```
:+1:

layus commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,316 @@
+
+    args = parser.parse_args()
+    print("automatically fixing dependencies for ELF files")
+    pprint.pprint(vars(args))
```
In the context of autoPatchelfHook it makes sense to keep this in a way or another. It allows to see what has been used in the end, wich is not obvious with all the addToSearchPath magic and other artifacts in hooks.

layus committed:
Apply suggestions from code review

Co-authored-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,304 @@
+            """)
+
+def main():
+
```
Nitpick:

```suggestion
```

aszlig commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,304 @@
+def main():
+
+    parser = argparse.ArgumentParser(prog="autoPatchelf.py",
+            description="""autoPatchelf tries as hard as possible to patch the provided\
+                    binary files by looking for compatible libraries in the provided paths.
+                    """)
```
Also nitpick, the following would avoid the unnecessary whitespace:

```suggestion
        description='Patch automatically discovered library dependencies into'
                    ' the provided executable.')
```

layus committed:
Fix issues discovered during tests

layus committed:
Apply suggestions from code review

Co-authored-by: aszlig <aszlig@redmoonstudios.org>

layus commented:
I fail to properly test this on androidenv. Accessing the 25.* tools version is not trivial.

layus committed:
fixup line wrapping

layus committed:
autoPatchelfHook: Improve compatibility with bash version

layus committed:
autoPatchelfHook: Fix symlink-reated issues

layus committed:
autoPatchelfHook: Revert dubious patchelf invocation test

layus committed:
autoPatchelfHook: Untangle the executable detection logic

layus committed:
fixup! autoPatchelfHook: Untangle the executable detection logic

Artturin commented:
Most reverse deps failed to build
Example log of 1password
https://termbin.com/fvs5
```log
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/h5h0j1cmfbngd64wp25ngw65zf0c0r1m-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
no Makefile, doing nothing
@nix { "action": "setPhase", "phase": "installPhase" }
installing
@nix { "action": "setPhase", "phase": "fixupPhase" }
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2
shrinking /nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2/bin/op
strip is /nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/bin/strip
stripping (with command strip and flags -S) in /nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2/bin 
patching script interpreter paths in /nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2
checking for references to /build/ in /nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2...
automatically fixing dependencies for ELF files
{'ignoreMissing': False,
 'libs': [PosixPath('/nix/store/cz8pz8h26z4xab7asqnpds3mq50kr2vy-auto-patchelf-py-hook/lib'),
          PosixPath('/nix/store/pjba17qx8vh7dln8m5vnkqdbbjahvg9b-patchelf-0.13/lib'),
          PosixPath('/nix/store/59jmzisg8fkm9c125fw384dqq1np602l-move-docs.sh/lib'),
          PosixPath('/nix/store/kxw6q8v6isaqjm702d71n2421cxamq68-make-symlinks-relative.sh/lib'),
          PosixPath('/nix/store/m54bmrhj6fqz8nds5zcj97w9s9bckc9v-compress-man-pages.sh/lib'),
          PosixPath('/nix/store/bkxq1nfi6grmww5756ynr1aph7w04lkk-strip.sh/lib'),
          PosixPath('/nix/store/bnj8d7mvbkg3vdb07yz74yhl3g107qq5-patch-shebangs.sh/lib'),
          PosixPath('/nix/store/cickvswrvann041nqxb0rxilc46svw1n-prune-libtool-files.sh/lib'),
          PosixPath('/nix/store/8zxndz5ag0p6s526c2xyllhk1nrn4c3i-audit-tmpdir.sh/lib'),
          PosixPath('/nix/store/c8n9kcdddp9np665xz6ri61b383nxvz8-move-systemd-user-units.sh/lib'),
          PosixPath('/nix/store/1i5y55x4b4m9qkx5dqbmr1r6bvrqbanw-multiple-outputs.sh/lib'),
          PosixPath('/nix/store/kd4xwxjpjxi71jkm6ka0np72if9rm3y0-move-sbin.sh/lib'),
          PosixPath('/nix/store/fyaryjvghbkpfnsyw97hb3lyb37s1pd6-move-lib64.sh/lib'),
          PosixPath('/nix/store/ngg1cv31c8c7bcm2n8ww4g06nq7s4zhm-set-source-date-epoch-to-latest.sh/lib'),
          PosixPath('/nix/store/wlwcf1nw2b21m4gghj70hbg1v7x53ld8-reproducible-builds.sh/lib'),
          PosixPath('/nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/lib'),
          PosixPath('/nix/store/v7w0pspwq8r2b7k2sndxq3db843z7xm5-binutils-wrapper-2.35.2/lib'),
          PosixPath('/nix/store/cz8pz8h26z4xab7asqnpds3mq50kr2vy-auto-patchelf-py-hook/lib'),
          PosixPath('/nix/store/pjba17qx8vh7dln8m5vnkqdbbjahvg9b-patchelf-0.13/lib'),
          PosixPath('/nix/store/59jmzisg8fkm9c125fw384dqq1np602l-move-docs.sh/lib'),
          PosixPath('/nix/store/kxw6q8v6isaqjm702d71n2421cxamq68-make-symlinks-relative.sh/lib'),
          PosixPath('/nix/store/m54bmrhj6fqz8nds5zcj97w9s9bckc9v-compress-man-pages.sh/lib'),
          PosixPath('/nix/store/bkxq1nfi6grmww5756ynr1aph7w04lkk-strip.sh/lib'),
          PosixPath('/nix/store/bnj8d7mvbkg3vdb07yz74yhl3g107qq5-patch-shebangs.sh/lib'),
          PosixPath('/nix/store/cickvswrvann041nqxb0rxilc46svw1n-prune-libtool-files.sh/lib'),
          PosixPath('/nix/store/8zxndz5ag0p6s526c2xyllhk1nrn4c3i-audit-tmpdir.sh/lib'),
          PosixPath('/nix/store/c8n9kcdddp9np665xz6ri61b383nxvz8-move-systemd-user-units.sh/lib'),
          PosixPath('/nix/store/1i5y55x4b4m9qkx5dqbmr1r6bvrqbanw-multiple-outputs.sh/lib'),
          PosixPath('/nix/store/kd4xwxjpjxi71jkm6ka0np72if9rm3y0-move-sbin.sh/lib'),
          PosixPath('/nix/store/fyaryjvghbkpfnsyw97hb3lyb37s1pd6-move-lib64.sh/lib'),
          PosixPath('/nix/store/ngg1cv31c8c7bcm2n8ww4g06nq7s4zhm-set-source-date-epoch-to-latest.sh/lib'),
          PosixPath('/nix/store/wlwcf1nw2b21m4gghj70hbg1v7x53ld8-reproducible-builds.sh/lib'),
          PosixPath('/nix/store/xiq6j4jsyj351p8q3yw9cg1hdqp9m685-gcc-wrapper-10.3.0/lib'),
          PosixPath('/nix/store/v7w0pspwq8r2b7k2sndxq3db843z7xm5-binutils-wrapper-2.35.2/lib')],
 'paths': [PosixPath('/nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2')],
 'recursive': True,
 'runtime_dependencies': []}
searching for dependencies of /nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2/bin/op
setting interpreter of /nix/store/fm04jy2csjav9wvwzcfr53s6wfbdp5kq-1password-1.12.2/bin/op
Traceback (most recent call last):
  File "/nix/store/lycra8danr1rkw20xp2igfj51bggj9a1-auto-patchelf.py", line 319, in <module>
    main()
  File "/nix/store/lycra8danr1rkw20xp2igfj51bggj9a1-auto-patchelf.py", line 296, in main
    autoPatchelf(
  File "/nix/store/lycra8danr1rkw20xp2igfj51bggj9a1-auto-patchelf.py", line 250, in autoPatchelf
    autoPatchelfFile(path, runtime_deps)
  File "/nix/store/lycra8danr1rkw20xp2igfj51bggj9a1-auto-patchelf.py", line 202, in autoPatchelfFile
    for dep in dependencies:
  File "/nix/store/lycra8danr1rkw20xp2igfj51bggj9a1-auto-patchelf.py", line 43, in get_dependencies
    for section in elf.iter_sections():
  File "/nix/store/mg6w3fh347i2axcc56kv7aswfqfphqyd-python3-3.9.6-env/lib/python3.9/site-packages/elftools/elf/elffile.py", line 133, in iter_sections
    yield self.get_section(i)
  File "/nix/store/mg6w3fh347i2axcc56kv7aswfqfphqyd-python3-3.9.6-env/lib/python3.9/site-packages/elftools/elf/elffile.py", line 112, in get_section
    section_header = self._get_section_header(n)
  File "/nix/store/mg6w3fh347i2axcc56kv7aswfqfphqyd-python3-3.9.6-env/lib/python3.9/site-packages/elftools/elf/elffile.py", line 526, in _get_section_header
    return struct_parse(
  File "/nix/store/mg6w3fh347i2axcc56kv7aswfqfphqyd-python3-3.9.6-env/lib/python3.9/site-packages/elftools/common/utils.py", line 39, in struct_parse
    stream.seek(stream_pos)
ValueError: seek of closed file
```

Report
https://termbin.com/1gr5

layus committed:
autoPatchelfHook: Fix invalid borrow issue

layus commented:
@Artturin yes, fixed, thanks. I used a generator that kept a reference to a stream that was closed before the actual use of the generator.

A this stage I am quite confident that we are reaching the end of WIP. nix-review manages to build everything, except for some rare cases like discordchatexporter-cli which fails now because I am more eager at finding the interpreter file. It did not fail before because the package contains no path to fix. I kept the failure as I think we should not litter the code with autoPatchelfHooks. That is also why I added a warning when nothing was patched. I would like to make it an error, but there are surprisingly quite a few packages that use autoPatchelfHook with nothing to patch. I think most of these use cases are invalid, but there are some valid corner cases.

layus commented:
@aszlig, @Artturin  Would you mind giving this one a second pass ? I think it matured well enough. At least it builds a lot a lot of things correctly in nix-review.

Attention points:
- I have changed the semantics of runtimeDirs to stop appending `/lib` at the end. It seems more flexible that way. Backward compatibility is preserved in the bash wrapper.
- I do not call patchelf for reading. It may cause some discrepancies with the old behavior but 1. I am confident enough that I replicated the exact same behavior and 2. it actually helped me untangle the exact semantics of the old hook.
- I added a warning when the hook does nothing. Should I make that louder ? Maybe add more empty lines around :-).
- Style, correctness, and error handling. (for example, when NIX_BINTOOLS is unset, you get an ugly error trace: 
  ```
    Traceback (most recent call last):
      File "/nix/store/bjj245b9pa7d4n9bdn758ks4m04rp9gg-auto-patchelf.py", line 311, in <module>
        nix_support = Path(os.environ['NIX_BINTOOLS']) / 'nix-support'
      File "/nix/store/k0z9n599k02hab8qjjp3ljw065iwjcvg-python3-3.9.6/lib/python3.9/os.py", line 679, in __getitem__
        raise KeyError(key) from None
    KeyError: 'NIX_BINTOOLS'
   ```
  Also, the phase start header is not printed, because it only occurs after that error. Should we move the header to the wrapper script ?

layus commented:
P.S. I am not against type annotations, but I did not manage to get `Tuple` to work for example.

layus committed:
autoPatchelfHook: Handle runtimeDependencies as the bare string it is

layus commented:
And another fix :tada:. Does this ever end ?

layus committed:
autoPatchelfHook: add bintools dependency

For the very rare cases where it is not included by default.

layus committed:
autoPatchelfHook: replace old hook with the rewrite

layus committed:
autoPatchelfHook: get rid of the old hook content

layus committed:
autoPatchelfHook: fix wrong ordering of debug info

layus committed:
autoPatchelfHook: persist extra search path across incovations

layus committed:
autoPatchelfHook: fix wrong usage of global variables

aszlig commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,332 @@
+    return elf.header["e_ident"]["EI_OSABI"]
+
+
+# Tests whether two OS ABIs are compatible, taking into account the
+# generally accepted compatibility of SVR4 ABI with other ABIs.
+def osabi_are_compatible(wanted, got):
```
```suggestion
def osabi_are_compatible(wanted, got):
    """
    Tests whether two OS ABIs are compatible, taking into account the
    generally accepted compatibility of SVR4 ABI with other ABIs.
    """
```

aszlig commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,332 @@
+                    'libraries in the provided paths.')
+    parser.add_argument(
+        "--ignore-missing",
+        dest="ignoreMissing",
```
Btw. you don't need those, since `argparse` by default will generate PEP-8 compatible attributes. In this case you'll get `ignore_missing`.

aszlig commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,332 @@
+        if not path.is_symlink() and path.is_file():
+            dependencies += autoPatchelfFile(path, runtime_deps)
+
+    # Print a summary of the missing dependencies at the end
+    for dep in dependencies:
+        if not dep.found:
+            print(f"auto-patchelf could not satisfy dependency {dep.name} wanted by {dep.file}")
+
+    any_missing = any(not dep.found for dep in dependencies)
+    if any_missing and not ignoreMissing:
```
```suggestion
    missing = [dep for dep in dependencies if not dep.found]
    
    # Print a summary of the missing dependencies at the end
    for dep in missing:
        print(f"auto-patchelf could not satisfy dependency {dep.name} wanted by {dep.file}")

    if missing and not ignore_missing:
```

aszlig reviewed (commented):
Mostly nitpicks from me, however I'm adding @svanderburg to the reviewers since he's built something similar for patching the Android SDK back then, which is why `autoPatchelfHook` can be used in a standalone way in order to support that use case. Since I'm not familiar with Android SDK packaging in nixpkgs, he's a better canidate for making sure that it still works after we merge this very pull request.

See also his [blog post](http://sandervanderburg.blogspot.com/2018/10/auto-patching-prebuilt-binary-software.html), particularly the `Follow up` section.

layus committed:
Update auto-patchelf.py

PEP8: ignoreMissing -> ignore_missing

layus committed:
Apply suggestions from code review

Co-authored-by: aszlig <aszlig@redmoonstudios.org>

layus committed:
autoPatchelfHook: remove imprecise and incorrect warning

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+                 'Add the missing dependencies to --libs or use --ignore-missing.')
+
+
+def main():
```
```suggestion
def main() -> None:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+soname_cache = defaultdict(list)
+
+
+def populate_cache(initial, recursive=False):
```
```suggestion
def populate_cache(initial: str, recursive: bool =False) -> None:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return wanted == got
+
+
+def glob(path, pattern, recursive):
```
```suggestion
def glob(path: str, pattern: str, recursive: bool) -> List[str]:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return elf.header["e_ident"]["EI_OSABI"]
+
+
+def osabi_are_compatible(wanted, got):
```
```suggestion
def osabi_are_compatible(wanted: Optional[str], got: Optional[str]) -> bool:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return elf.get_machine_arch()
+
+
+def get_osabi(elf):
```
```suggestion
def get_osabi(elf: ElfFile) -> str:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return []
+
+
+def get_arch(elf):
```
```suggestion
def get_arch(elf: ELFFile) -> str:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return dependencies
+
+
+def get_rpath(elf):
```
```suggestion
def get_rpath(elf: ELFFile) -> List[str]:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return bool(elf.get_section_by_name(".interp"))
+
+
+def get_dependencies(elf):
```
```suggestion
def get_dependencies(elf: ELFFile) -> List[str]:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+        yield ELFFile(stream)
+
+
+def is_static_executable(elf):
```
```suggestion
def is_static_executable(elf: ELFFile) -> bool:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+            and not elf.get_section_by_name(".interp"))
+
+
+def is_dynamic_executable(elf):
```
```suggestion
def is_dynamic_executable(elf: ELFFile) -> bool:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+
+
+@contextmanager
+def open_elf(path):
```
```suggestion
def open_elf(path: str) -> Iterator[IO[byte]]:
```

aszlig replied:
The `path` argument actually has type `Path`

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return dependencies
+
+
+def autoPatchelf(
+        pathsToPatch,
+        libDirs,
+        runtime_deps,
+        recursive=True,
+        ignoreMissing=False):
```
```suggestion
def autoPatchelf(
        pathsToPatch,
        libDirs,
        runtime_deps,
        recursive: bool =True,
        ignoreMissing: bool =False) -> None:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+                pass
+
+
+def findDependency(soname, soarch, soabi):
```
```suggestion
def findDependency(soname: str, soarch: str, soabi: str) -> Optional[str]:
```

Mic92 replied:
```suggestion
def find_dependency(soname, soarch, soabi):
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    found: bool = False     # Whether it was found somewhere
+
+
+def autoPatchelfFile(path: Path, runtime_deps: list[Path]) -> list[Dependency]:
```
```suggestion
def auto_patchelf_file(path: Path, runtime_deps: list[Path]) -> list[Dependency]:
```

Mic92 commented on pkgs/build-support/setup-hooks/auto-patchelf.py:
```diff
@@ -0,0 +1,328 @@
+    return dependencies
+
+
+def autoPatchelf(
```
```suggestion
def auto_patchelf(
```

Mic92 commented:
I guessed all the types, so some of them might be wrong.

Mic92 commented:
It would be great if you could also for variables stick to `snake_case` instead of mixing `camelCase` and `snake_case`.

layus committed:
Apply explicit types from code review

Co-authored-by: Jörg Thalheim <Mic92@users.noreply.github.com>

layus committed:
Complement and polish types and snake_casing

layus commented:
@Mic92 Thanks for review and edits. I applied your suggestions and made everything type and snake_case consistent :-).

Mic92 commented:
I did a nixpkgs-review, which did not completely finish for unrelated reasons, but I could not spot regressions in the failing builds.

Mic92 commented:
@aszlig do you think this is good to go?

Artturin commented:
Target staging?

layus commented:
Rebased, retargeted. Sorry for the mess of review requests that created.

DavHau commented:
For maximum performance it would be cool if patching the individual files would be handled by a thread pool or process pool. But I am happy to do this in a follow up PR.

FRidh commented:
If we do run into issues with say the sdk and it is not easily fixable we can always revert.

layus commented:
Thanks everyone for the review process. It has been a pleasant experience.
